Prefix range: simplify

dhalperi opened the pull request:
Three changes to simplify use of PrefixRange:

* simplify construction PrefixSpace construction - relax `Set<PrefixRange>` to `Iterable<PrefixRange>`, add a varargs version. 
* introduce `DEFAULT_PREFIX_ONLY`, which is the same as `PrefixRange(0.0.0.0/0, [0, 0])`
* introduce `PrefixRange.moreSpecificThan(prefix)`, which is the same as `PrefixRange(p, [p.length+1, 32])`.

progwriter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1283#-:-LBXqgf78QwKC95uenrn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

arifogel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one suggested refactoring.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java, line 49 at r1](https://reviewable.io/reviews/batfish/batfish/1283#-LBYd8MR-_2rTrne81Hu:-LBYd8MSEJ_uQU8sTfvs:b-e56enk) ([raw file](https://github.com/batfish/batfish/blob/4287c564b26c4852106b3e14135af461e65eac06/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java#L49)):*
> ```Java
>   public static PrefixRange moreSpecificThan(Prefix prefix) {
>     return new PrefixRange(
>         prefix, new SubRange(prefix.getPrefixLength() + 1, Prefix.MAX_PREFIX_LENGTH));
> ```

What happens if `prefix.getPrefixLength == Prefix.MAX_PREFIX_LENGTH`?

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 1367 at r1](https://reviewable.io/reviews/batfish/batfish/1283#-LBYdcTc53ndBLWqh-YA:-LBYdcTc53ndBLWqh-YB:b9w312f) ([raw file](https://github.com/batfish/batfish/blob/4287c564b26c4852106b3e14135af461e65eac06/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L1367)):*
> ```Java
>         new MatchPrefixSet(
>             new DestinationNetwork(),
>             new ExplicitPrefixSet(new PrefixSpace(PrefixRange.DEFAULT_PREFIX_ONLY)));
> ```

Maybe also make `new ExplicitPrefixSet(new PrefixSpace(PrefixRange.DEFAULT_PREFIX_ONLY))` a constant in `ExplicitPrefixSet`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1283#-:-LBYe37h6x44FZ-oBJs5:b-yyq8ls)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: 6 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java, line 49 at r1](https://reviewable.io/reviews/batfish/batfish/1283#-LBYd8MR-_2rTrne81Hu:-LBYgQbA-rS5Cg4vSPif:b-r3ksym) ([raw file](https://github.com/batfish/batfish/blob/4287c564b26c4852106b3e14135af461e65eac06/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/PrefixRange.java#L49)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

What happens if `prefix.getPrefixLength == Prefix.MAX_PREFIX_LENGTH`?
</blockquote></details>

This generally shouldn't happen given correct user input of course, but the actual behavior is I think the desired behavior: the PrefixRange is considered empty -- or at least it doesn't contain the /32, which I think is roughly as close as we get to empty.

(Added a test)

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 1367 at r1](https://reviewable.io/reviews/batfish/batfish/1283#-LBYdcTc53ndBLWqh-YA:-LBYkrHtA8ny_TPThMi-:b-p0d5c6) ([raw file](https://github.com/batfish/batfish/blob/4287c564b26c4852106b3e14135af461e65eac06/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L1367)):*
<details><summary><i>Previously, arifogel (Ari Fogel) wrote…</i></summary><blockquote>

Maybe also make `new ExplicitPrefixSet(new PrefixSpace(PrefixRange.DEFAULT_PREFIX_ONLY))` a constant in `ExplicitPrefixSet`?
</blockquote></details>

Decided to move this out of the PrefixRange class and into this file, where it makes more sense. In doing so, I simplified a bunch more code and fixed a minor bug.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1283)*
<!-- Sent from Reviewable.io -->

arifogel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1283#-:-LBZGnJo0MKbW5REky3E:bnfp4nl)*
<!-- Sent from Reviewable.io -->
